service/matching/matching_engine_test.go

```diff
@@ -2761,6 +2762,321 @@ func (s *matchingEngineSuite) TestUpdatePhysicalTaskQueueGauge_BuildID() {
+	workflowType, workflowExecution := s.generateWorkflowExecution()
+	taskQueue, ptq := s.createTQAndPTQForBacklogTests()
+
+	const taskCount = 100
+	s.addWorkflowTasks(false, 1, taskCount, taskQueue, workflowExecution, nil)
+	s.EqualValues(taskCount, s.taskManager.getTaskCount(ptq))
+
+	// Extracting the tlMgr for validating approximateBacklogCounter
+	tlMgr := s.getPhysicalTaskQueueManagerImpl(ptq)
+	s.EqualValues(tlMgr.backlogMgr.db.getApproximateBacklogCount(), int64(taskCount))
+
+	s.mockHistoryWhilePolling(workflowType)
+
+	s.pollWorkflowTasks(false, 1, taskCount, tlMgr, taskQueue, nil)
+	// All tasks should have been completed
+	s.EqualValues(tlMgr.backlogMgr.db.getApproximateBacklogCount(), int64(0))
+}
+
+// TestResettingBacklogCounter tests the scenario where approximateBacklogCounter undercounts and resets it accordingly
+func (s *matchingEngineSuite) TestCheckPreventionOfNegativeBacklogCounter() {
+	// The backlogCounter is decreased only when CompleteTask is called on the AckManager.
+	// We shall manually remove the task from the testTaskManager and then call AckManager.CompleteTask(TaskID) to see the resetting take place
+
+	// TODO Shivam - This should fail with the current implementation. Have a skip or something.
+
+	s.matchingEngine.config.LongPollExpirationInterval = dynamicconfig.GetDurationPropertyFnFilteredByTaskQueueInfo(10 * time.Millisecond)
+	_, workflowExecution := s.generateWorkflowExecution()
+
+	const taskCount = 1
+	const taskID = 11
+
+	taskQueue, ptq := s.createTQAndPTQForBacklogTests()
+	s.addWorkflowTasks(false, 1, taskCount, taskQueue, workflowExecution, nil)
+	s.EqualValues(taskCount, s.taskManager.getTaskCount(ptq))
+
+	// validating the approximateBacklogCounter
+	tlMgr := s.getPhysicalTaskQueueManagerImpl(ptq)
+	s.EqualValues(tlMgr.backlogMgr.db.getApproximateBacklogCount(), int64(1))
+
+	// Remove the task from testTaskManager but not from db/AckManager
+	request := &persistence.CompleteTasksLessThanRequest{
+		NamespaceID:        namespaceId,
+		TaskQueueName:      taskQueue.Name,
+		TaskType:           enumspb.TASK_QUEUE_TYPE_WORKFLOW,
+		ExclusiveMaxTaskID: taskID + 1,
+		Limit:              100,
+	}
+	_, err := s.taskManager.CompleteTasksLessThan(context.Background(), request)
+	s.NoError(err)
+	// Decreasing the backlogCounter since a task has been removed in theory
+	tlMgr.backlogMgr.db.updateApproximateBacklogCount(int64(-1))
+
+	// overriding DB's logger to catch ERROR's
+	mockLogger := log.NewMockLogger(s.controller)
+	mockLogger.EXPECT().Error("ApproximateBacklogCounter could have under-counted. This should never happen",
+		tag.WorkerBuildId(tlMgr.backlogMgr.db.queue.BuildId()), tag.WorkerBuildId(tlMgr.backlogMgr.db.queue.Partition().NamespaceId().String()))
+	tlMgr.backlogMgr.db.logger = mockLogger
+	_, ok := tlMgr.backlogMgr.db.logger.(*log.MockLogger)
+	s.Equal(ok, true)
```

Comment:
Could do; this test anyways needs refactoring since I believe it was developed a different idea in mind (as discussed async).

I will be refactoring this to check the `reset` logic where the counter will be an over-estimate and shall be set to 0.